src/java.base/share/classes/java/lang/Integer.java

```diff
@@ -516,52 +519,65 @@ public static String toUnsignedString(int i) {
+            return parseInt0(s, radix);
         }
-
-        int len = s.length();
-        if (len == 0) {
-            throw NumberFormatException.forInputString("", radix);
+        int result = 0, c, c1, digit;
+        boolean inRange;
+        int neg;
+        if ((neg = (c = value[0]) - '-') != 0
+                && neg + 2 != 0 // firstChar != '+'
+        ) {
+            if (inRange = isDigit(c)) {
+                result = '0' - c;
+            }
+        } else {
+            inRange = len != 1;
         }
-        int digit = ~0xFF;
-        int i = 0;
-        char firstChar = s.charAt(i++);
-        if (firstChar != '-' && firstChar != '+') {
-            digit = digit(firstChar, radix);
+        int limit = MIN_VALUE + (neg != 0 ? 1 : 0);
+        int i = 1;
+        while (inRange
+                && i + 1 < len
+                && isDigit((c = value[i]))
+                && isDigit(c1 = value[i + 1])
+        ) {
+            digit = c * 10 + c1 - 528; // 528 = 48 * 11 = '0' * 10 + '0'
+            // max digits is 19, no need to check inRange (result == MULT_MIN_100 && digit <= (MULT_MIN_100 * 100 - limit))
+            if (inRange = (result >= MULT_MIN_100)) {
+                result = result * 100 - digit;
+                i += 2;
+            }
         }
-        if (digit >= 0 || digit == ~0xFF && len > 1) {
-            int limit = firstChar != '-' ? MIN_VALUE + 1 : MIN_VALUE;
-            int multmin = limit / radix;
-            int result = -(digit & 0xFF);
-            boolean inRange = true;
-            /* Accumulating negatively avoids surprises near MAX_VALUE */
-            while (i < len && (digit = digit(s.charAt(i++), radix)) >= 0
-                    && (inRange = result > multmin
-                        || result == multmin && digit <= radix * multmin - limit)) {
-                result = radix * result - digit;
+        if (inRange && result <= 0) {
+            if (i + 1 == len && isDigit((c = value[i]))) {
+                digit = c - '0';
+                if (result > MULT_MIN_10 || (result == MULT_MIN_10 && digit <= (MULT_MIN_10 * 10 - limit))) {
+                    result = result * 10 - digit;
+                    i++;
+                }
             }
-            if (inRange && i == len && digit >= 0) {
-                return firstChar != '-' ? -result : result;
+            if (i == len) {
+                return neg != 0 ? -result : result;
             }
         }
-        throw NumberFormatException.forInputString(s, radix);
+        throw NumberFormatException.forInputString(s);
+    }
+
+    private static int parseInt0(String s, int radix) {
+        if (s == null) {
+            throw NumberFormatException.emptyInput();
+        }
+        int len;
+        if ((len = s.length()) == 0) {
+            throw NumberFormatException.forInputString(s);
+        }
+        return parseInt(s, 0, len, radix);
```

Comment:
I  think this call will result in different exception messages in some cases. Consolidating the exception messages between the String and the CharSequence methods would likely allow for even more code simplification,  but currently it appears there was effort to preserve the exact message text in exceptions.

Reply:
If we want to keep the original error message completely, we need to keep the original implementation, which will lead to code duplication.